aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -84,33 +94,52 @@ struct DispatchKeyExtractor final {
         }
       }
     }
-    return dispatchKeySetToDispatchKey_(ts);
+    return dispatchKeySetToDispatchKey_(nonFallthroughMask, ts);
   }
 
   template<class... Args>
-  c10::optional<DispatchKey> getDispatchKeyUnboxed(const Args&... args) const {
+  c10::optional<DispatchKey> getDispatchKeyUnboxed(DispatchKeySet nonFallthroughMask, const Args&... args) const {
     auto key_set = detail::multi_dispatch_key_set(args...);
-    return dispatchKeySetToDispatchKey_(key_set);
+    return dispatchKeySetToDispatchKey_(nonFallthroughMask, key_set);
   }
 
+  // Used by DispatchTable to maintain the fallthrough invariant, see
+  // docs on nonFallthroughKernels_
+  void setIsOperatorOverridden(DispatchKey k, bool is_overridden);
+
 private:
-  static c10::optional<DispatchKey> dispatchKeySetToDispatchKey_(const DispatchKeySet& keySet) {
+  c10::optional<DispatchKey> dispatchKeySetToDispatchKey_(DispatchKeySet nonFallthroughMask, const DispatchKeySet& keySet) const {
     if (C10_UNLIKELY(keySet.empty())) {
       return c10::nullopt;
     }
 
-    return impl::dispatchTypeId(keySet);
+    return impl::dispatchTypeId(keySet, nonFallthroughMask | perOperatorOverriddenKernels_);
   }
 
   explicit DispatchKeyExtractor(size_t num_args)
-  : num_args_(num_args) {}
+  : num_args_(num_args)
+  , perOperatorOverriddenKernels_() {}
 
   // this is caching the index so we don't have to parse the schema inputs
   // again and again for each dispatcher lookup.
   // num_args_ is allowed to be zero; that just means you must do the
   // fallthrough
   // TODO: a potential optimization is to store a bitfield of arg locations,
   size_t num_args_;
+
+  // We must NOT respect the passed in nonFallthroughMask if an operator has
+  // specifically overridden the backend, since that means we've opted to
+  // not fallthrough and instead apply some specific behavior (which we
+  // must dispatch to).  For now, we assume that operators NEVER are the
+  // fallthrough kernel (see https://github.com/pytorch/pytorch/issues/32454)
+  // which means we can just unconditionally fill in the mask when the
+  // operator tells us to.
+  //
+  // This scheme doesn't work if you want to also apply fallthrough on a
+  // per-op basis, but while we could fix this by maintaining a second
+  // DispatchKeySet, it doesn't seem that there is any actual use case,
+  // so we are deferring it for 32454.
+  DispatchKeySet perOperatorOverriddenKernels_;
```

Comment:
I find the name a bit confusing since this is not strictly about overriding backend fallbacks, this is also set if a backend doesn't even have a fallback but the operator has a kernel for this key. Maybe a better name is `setOperatorHasKernelForBackend` and `operatorHasKernelForBackends_`.

After that naming change, the comment doesn't need to be at the member variable anymore because the member variable is self-explanatory, so the comment can be moved to where the variable is used, i.e. into dispatchKeySetToDispatchKey_. That's probably where a reader of the code would wonder why you're or-ing this mask.